### ISSUE TITLE
Parallel reading of neuroelectro.dir, bak, cache still problematic on big runs.

### ISSUE THREAD

**Issue by russelljjarvis** [scidash/sciunit#63]
On bigger runs, the probability of a read not succeeding is greater. In theory parallel read should be fine, but there is something intrinsically not parallel friendly about the neuroelectro caching methods.

Parallel reading of neuroelectro.dir, bak, cache still problematic on big runs.

Ideally tests from get_neab should be pickable then it can be performed just once on only one CPU and then pushed to all the others. In practice though get_neab.tests is still not picklable.

This is my most current version of the bug. At least it demonstrates graceful failure.
```python
NeuroElectroError("NeuroElectro.org appears to be down.")
```

I think I disabled writing the pickle files together so it tries to download new files every time, making the program more amenable to web server errors.

Parallel writes are always a problem. 

```python
---------------------------------------------------------------------------gaierror                                  Traceback (most recent call last)/opt/conda/lib/python3.5/urllib/request.py in do_open(self, http_class, req, **http_conn_args)
   1253             try:
-> 1254                 h.request(req.get_method(), req.selector, req.data, headers)
   1255             except OSError as err: # timeout error
/opt/conda/lib/python3.5/http/client.py in request(self, method, url, body, headers)
   1105         """Send a complete request to the server."""
-> 1106         self._send_request(method, url, body, headers)
   1107 
/opt/conda/lib/python3.5/http/client.py in _send_request(self, method, url, body, headers)
   1150             body = _encode(body, 'body')
-> 1151         self.endheaders(body)
   1152 
/opt/conda/lib/python3.5/http/client.py in endheaders(self, message_body)
   1101             raise CannotSendHeader()
-> 1102         self._send_output(message_body)
   1103 
/opt/conda/lib/python3.5/http/client.py in _send_output(self, message_body)
    933 
--> 934         self.send(msg)
    935         if message_body is not None:
/opt/conda/lib/python3.5/http/client.py in send(self, data)
    876             if self.auto_open:
--> 877                 self.connect()
    878             else:
/opt/conda/lib/python3.5/http/client.py in connect(self)
    848         self.sock = self._create_connection(
--> 849             (self.host,self.port), self.timeout, self.source_address)
    850         self.sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
/opt/conda/lib/python3.5/socket.py in create_connection(address, timeout, source_address)
    692     err = None
--> 693     for res in getaddrinfo(host, port, 0, SOCK_STREAM):
    694         af, socktype, proto, canonname, sa = res
/opt/conda/lib/python3.5/socket.py in getaddrinfo(host, port, family, type, proto, flags)
    731     addrlist = []
--> 732     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
    733         af, socktype, proto, canonname, sa = res
gaierror: [Errno -2] Name or service not known
During handling of the above exception, another exception occurred:
URLError                                  Traceback (most recent call last)~/neuronunit/neuronunit/neuroelectro.py in get_json(self, params, quiet)
    149         try:
--> 150             url_result = urlopen(url,None,3) # Get the page.
    151             html = url_result.read() # Read out the HTML (actually JSON)
/opt/conda/lib/python3.5/urllib/request.py in urlopen(url, data, timeout, cafile, capath, cadefault, context)
    162         opener = _opener
--> 163     return opener.open(url, data, timeout)
    164 
/opt/conda/lib/python3.5/urllib/request.py in open(self, fullurl, data, timeout)
    465 
--> 466         response = self._open(req, data)
    467 
/opt/conda/lib/python3.5/urllib/request.py in _open(self, req, data)
    483         result = self._call_chain(self.handle_open, protocol, protocol +
--> 484                                   '_open', req)
    485         if result:
/opt/conda/lib/python3.5/urllib/request.py in _call_chain(self, chain, kind, meth_name, *args)
    443             func = getattr(handler, meth_name)
--> 444             result = func(*args)
    445             if result is not None:
/opt/conda/lib/python3.5/urllib/request.py in http_open(self, req)
   1281     def http_open(self, req):
-> 1282         return self.do_open(http.client.HTTPConnection, req)
   1283 
/opt/conda/lib/python3.5/urllib/request.py in do_open(self, http_class, req, **http_conn_args)
   1255             except OSError as err: # timeout error
-> 1256                 raise URLError(err)
   1257             r = h.getresponse()
URLError: <urlopen error [Errno -2] Name or service not known>
During handling of the above exception, another exception occurred:
AttributeError                            Traceback (most recent call last)~/neuronunit/neuronunit/neuroelectro.py in get_json(self, params, quiet)
    153             try:
--> 154                 html = e.read().decode('utf-8')
    155                 self.json_object = json.loads(html)
AttributeError: 'URLError' object has no attribute 'read'
During handling of the above exception, another exception occurred:
NeuroElectroError                         Traceback (most recent call last)<string> in <module>()
/opt/conda/lib/python3.5/site-packages/ipyparallel/client/remotefunction.py in <lambda>(f, *sequences)
    248             if _mapping:
    249                 if sys.version_info[0] >= 3:
--> 250                     f = lambda f, *sequences: list(map(f, *sequences))
    251                 else:
    252                     f = map
~/neuronunit/neuronunit/plottools.py in dtc_to_plotting(dtc)
    540     dtc.vm1 = None
    541     from neuronunit.models.reduced import ReducedModel
--> 542     from neuronunit.optimization.get_neab import tests as T
    543     from neuronunit.optimization import get_neab
    544     from neuronunit.optimization import evaluate_as_module
~/neuronunit/neuronunit/optimization/get_neab.py in <module>()
     58         #use of the variable 'neuron' in this conext conflicts with the module name 'neuron'
     59         #at the moment it doesn't seem to matter as neuron is encapsulated in a class, but this could cause problems in the future.
---> 60         observation = cls.neuroelectro_summary_observation(neuron)
     61         tests += [cls(observation)]
     62 
~/neuronunit/neuronunit/tests/base.py in neuroelectro_summary_observation(cls, neuron)
    104                                                      # NeuroElectro ontology.
    105             )
--> 106         reference_data.get_values(quiet=not cls.verbose) # Get and verify summary data
    107                                     # from neuroelectro.org.
    108 
~/neuronunit/neuronunit/neuroelectro.py in get_values(self, params, quiet)
    262     def get_values(self, params=None, quiet=False):
    263         data = super(NeuroElectroSummary, self).get_values(params=params,
--> 264                                                            quiet=quiet)
    265         if data:
    266             self.neuron.name = data['n']['name']
~/neuronunit/neuronunit/neuroelectro.py in get_values(self, params, quiet)
    188             self.json_object = json.loads(db[identifier])
    189         else:
--> 190             self.get_json(params=params, quiet=quiet)
    191             if DUMP:
    192                 db[identifier] = json.dumps(self.json_object)
~/neuronunit/neuronunit/neuroelectro.py in get_json(self, params, quiet)
    158             except:
    159                 if hasattr(e,'reason'):
--> 160                     raise NeuroElectroError(e.reason)
    161             raise NeuroElectroError("NeuroElectro.org appears to be down.")
    162             #print "Using fake data for now."
NeuroElectroError: [Errno -2] Name or service not known
```

**Comment by rgerkin** [comment]
Are you able to duplicate the failure using some very simple test that involves reading from a shelved (the current neuroelectro cache implementation) file several thousand times, with a parallel read eventually failing?  If so, we could modify that test with different caching implementations (e.g. various pickle libraries, databases, json, etc.) and see if one of them does not fail, and then switch to that.  

Could you let me know what about get_neab.tests is not picklable, e.g. what is the error (can you make a simple formula to reproduce this)?  Maybe I can get it working.

**Comment by russelljjarvis** [comment]
Okay I will try what you have suggested here.

**Comment by russelljjarvis** [comment]
I am finding that neuroelectro the website is down (confirmed both programatically and by pointing my browswer at the URL), the state of neuroelectro being down has motivated me to work on this issue, since at the moment I can't develop unless I can load the pickled version of the file, and I think it would be neater if it was read by one CPU and transmitted to the others.

While I have not yet made a formal unit test, below is a minimal working example that demonstrates the problem. 


```
 from ipyparallel import depend, require, dependent
 import ipyparallel as ipp
rc = ipp.Client(profile='default')
 rc[:].use_cloudpickle()
 dview = rc[:]
In [2]: tests2017-11-30 00:34:29.927 [IPClusterStart] Engines appear to have started successfully
In [2]: tests = pickle.load(open('neuroelectro.pickle','rb'))
```
```
In [8]: tests
Out[8]: 
[<neuronunit.tests.fi.RheobaseTestP at 0x7f6609235828>,
 <neuronunit.tests.passive.InputResistanceTest at 0x7f65db544828>,
 <neuronunit.tests.passive.TimeConstantTest at 0x7f65db544cc0>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f65db544d30>,
 <neuronunit.tests.passive.RestingPotentialTest at 0x7f65dc4db240>,
 <neuronunit.tests.waveform.InjectedCurrentAPWidthTest at 0x7f65db544e10>,
 <neuronunit.tests.waveform.InjectedCurrentAPAmplitudeTest at 0x7f65db544e48>,
 <neuronunit.tests.waveform.InjectedCurrentAPThresholdTest at 0x7f65db544e80>]

```
test_dic = {}
In [9]: for t in tests: test_dic[t]
```
```
In [10]: for t in tests: test_dic[t]= t
```
In [11]: dview.push(test_dic)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-16-85e9f055c2cb> in <module>()
----> 1 dview.push(test_dic)

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in push(self, ns, targets, block, track)
    700         if not isinstance(ns, dict):
    701             raise TypeError("Must be a dict, not %s"%type(ns))
--> 702         return self._really_apply(util._push, kwargs=ns, block=block, track=track, targets=targets)
    703 
    704     def get(self, key_s):

<decorator-gen-131> in _really_apply(self, f, args, kwargs, targets, block, track)

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in sync_results(f, self, *args, **kwargs)
     48     self._in_sync_results = True
     49     try:
---> 50         ret = f(self, *args, **kwargs)
     51     finally:
     52         self._in_sync_results = False

<decorator-gen-130> in _really_apply(self, f, args, kwargs, targets, block, track)

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in save_ids(f, self, *args, **kwargs)
     33     n_previous = len(self.client.history)
     34     try:
---> 35         ret = f(self, *args, **kwargs)
     36     finally:
     37         nmsgs = len(self.client.history) - n_previous

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in _really_apply(self, f, args, kwargs, targets, block, track)
    555         for ident in _idents:
    556             future = self.client.send_apply_request(self._socket, f, args, kwargs, track=track,
--> 557                                     ident=ident)
    558             futures.append(future)
    559         if track:

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/client.py in send_apply_request(self, socket, f, args, kwargs, metadata, track, ident)
   1393         bufs = serialize.pack_apply_message(f, args, kwargs,
   1394             buffer_threshold=self.session.buffer_threshold,
-> 1395             item_threshold=self.session.item_threshold,
   1396         )
   1397 

/opt/conda/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py in pack_apply_message(f, args, kwargs, buffer_threshold, item_threshold)
    166         serialize_object(arg, buffer_threshold, item_threshold) for arg in args))
    167 
--> 168     kw_keys = sorted(kwargs.keys())
    169     kwarg_bufs = list(chain.from_iterable(
    170         serialize_object(kwargs[key], buffer_threshold, item_threshold) for key in kw_keys))

TypeError: unorderable types: RheobaseTestP() < InjectedCurrentAPWidthTest()
```
```
In [17]: test_dic = {}

In [18]: for t in tests[1:-1]: test_dic[t] = t

In [19]: test_dic
Out[19]: 
{<neuronunit.tests.passive.CapacitanceTest at 0x7f65db544d30>: <neuronunit.tests.passive.CapacitanceTest at 0x7f65db544d30>,
 <neuronunit.tests.waveform.InjectedCurrentAPAmplitudeTest at 0x7f65db544e48>: <neuronunit.tests.waveform.InjectedCurrentAPAmplitudeTest at 0x7f65db544e48>,
 <neuronunit.tests.waveform.InjectedCurrentAPWidthTest at 0x7f65db544e10>: <neuronunit.tests.waveform.InjectedCurrentAPWidthTest at 0x7f65db544e10>,
 <neuronunit.tests.passive.InputResistanceTest at 0x7f65db544828>: <neuronunit.tests.passive.InputResistanceTest at 0x7f65db544828>,
 <neuronunit.tests.passive.RestingPotentialTest at 0x7f65dc4db240>: <neuronunit.tests.passive.RestingPotentialTest at 0x7f65dc4db240>,
 <neuronunit.tests.passive.TimeConstantTest at 0x7f65db544cc0>: <neuronunit.tests.passive.TimeConstantTest at 0x7f65db544cc0>}

In [20]: dview.push(test_dic)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-20-85e9f055c2cb> in <module>()
----> 1 dview.push(test_dic)

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in push(self, ns, targets, block, track)
    700         if not isinstance(ns, dict):
    701             raise TypeError("Must be a dict, not %s"%type(ns))
--> 702         return self._really_apply(util._push, kwargs=ns, block=block, track=track, targets=targets)
    703 
    704     def get(self, key_s):

<decorator-gen-131> in _really_apply(self, f, args, kwargs, targets, block, track)

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in sync_results(f, self, *args, **kwargs)
     48     self._in_sync_results = True
     49     try:
---> 50         ret = f(self, *args, **kwargs)
     51     finally:
     52         self._in_sync_results = False

<decorator-gen-130> in _really_apply(self, f, args, kwargs, targets, block, track)

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in save_ids(f, self, *args, **kwargs)
     33     n_previous = len(self.client.history)
     34     try:
---> 35         ret = f(self, *args, **kwargs)
     36     finally:
     37         nmsgs = len(self.client.history) - n_previous

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/view.py in _really_apply(self, f, args, kwargs, targets, block, track)
    555         for ident in _idents:
    556             future = self.client.send_apply_request(self._socket, f, args, kwargs, track=track,
--> 557                                     ident=ident)
    558             futures.append(future)
    559         if track:

/opt/conda/lib/python3.5/site-packages/ipyparallel/client/client.py in send_apply_request(self, socket, f, args, kwargs, metadata, track, ident)
   1393         bufs = serialize.pack_apply_message(f, args, kwargs,
   1394             buffer_threshold=self.session.buffer_threshold,
-> 1395             item_threshold=self.session.item_threshold,
   1396         )
   1397 

/opt/conda/lib/python3.5/site-packages/ipyparallel/serialize/serialize.py in pack_apply_message(f, args, kwargs, buffer_threshold, item_threshold)
    166         serialize_object(arg, buffer_threshold, item_threshold) for arg in args))
    167 
--> 168     kw_keys = sorted(kwargs.keys())
    169     kwarg_bufs = list(chain.from_iterable(
    170         serialize_object(kwargs[key], buffer_threshold, item_threshold) for key in kw_keys))

TypeError: unorderable types: InputResistanceTest() < InjectedCurrentAPWidthTest()
```

**Comment by russelljjarvis** [comment]
So although I could push to workers okay, I found that I couldn't pull even really simple dictionaries. I factored cloud pickle out, and noticed that, that can make a critical difference, since we are no longer trying to pickle NU models, we probably no longer need cloud pickle anyway. 

```
import ipyparallel as ipp
rc = ipp.Client(profile='default')
rc[:].use_cloudpickle()
dview = rc[:]
```
```
In [53]: dview.push(dict(a=1.03234,b=3453))
Out[53]: <AsyncResult: _push>

In [54]: dview.pull('a')
Out[54]: <AsyncResult: _pull>
In [55]: engine.0.error:
    ---------------------------------------------------------------------------
    NameError                                 Traceback (most recent call last)
    <string> in <module>()

    /opt/conda/lib/python3.5/site-packages/ipyparallel/util.py in _pull(keys)
    270         return [eval(key, globals()) for key in keys]
    271     else:
--> 272         return eval(keys, globals())
    273 
    274 @interactive

    <string> in <module>()

    NameError: name 'a' is not defined
engine.0.error:
    ---------------------------------------------------------------------------
    NameError                                 Traceback (most recent call last)
    <string> in <module>()

    /opt/conda/lib/python3.5/site-packages/ipyparallel/util.py in _pull(keys)
    270         return [eval(key, globals()) for key in keys]
    271     else:
--> 272         return eval(keys, globals())
    273 
    274 @interactive

    <string> in <module>()

    NameError: name 'a' is not defined
engine.1.error:

```

```
In [44]: 
    ...: dview.pull('RheobaseTestP')
    ...: 
    ...: 
    ...: 
Out[44]: <AsyncResult: _pull>

In [45]: engine.5.error:
    ---------------------------------------------------------------------------
    NameError                                 Traceback (most recent call last)
    <string> in <module>()

    /opt/conda/lib/python3.5/site-packages/ipyparallel/util.py in _pull(keys)
    270         return [eval(key, globals()) for key in keys]
    271     else:
--> 272         return eval(keys, globals())
    273 
    274 @interactive

    <string> in <module>()

    NameError: name 'RheobaseTestP' is not defined
```
exit and restart ipython:
```
import ipyparallel as ipp
rc = ipp.Client(profile='default')
dview = rc[:]
```
```
In [19]: dview.push(test_dic)
Out[19]: [None, None, None, None, None, None, None, None]
```
```
In [20]: dview.get('CapacitanceTest')
Out[20]: 
[<neuronunit.tests.passive.CapacitanceTest at 0x7f31f4ee2e10>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e6aa90>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e6a4a8>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e6f0f0>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e6f710>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e749b0>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e74390>,
 <neuronunit.tests.passive.CapacitanceTest at 0x7f31f4e6fd30>]
```

**Comment by rgerkin** [comment]
@russelljjarvis What is the current status of this?

**Comment by russelljjarvis** [comment]
This issue consists of two related sub-issues.


Neuroelectro reading may still be a problem.

Issue 1.
 For a while, calls to the neuroelectro API were giving 'server-down' exceptions. I don't know if that has changed. To make the ipython notebook that I showed you in the lab I disabled web downloading, and went straight to reading from cached values, as I have no control over the neuroelectro server. Enough time has elapsed that it would be worth checking server up status.

Issue 2.
In my code base, I made it such that reading the neuro_electro cache from the get_neab.py file occurred minimally. Calling get_neab.py once would be ideal, but I could not implement a single call with ipyparallel (that's also what the discussion above was about), trying to read only one version of neuronunit test suite object from CPU0 and to share it around with the workers (CPU1,...,CPUN).

So far ipyparallel cannot support pushing and pulling of the test suite, at the same time as using cloud pickle. It's raised as an issue on the ipyparallel GH issues page. So far there is no solution and 'aliasing' of variables inside the ipyparallel package has been blamed.

I suggest that we move to dask distributed instead in the future, as I think it is a more mature project, and it does not require a clumsy invocation like:
os.system('ipcluster -n4 start') etc.
